### PR TITLE
Migrate to GitHub Actions

### LINKED ISSUE #6787: Migrate CI to GitHub Actions
Due to changes in the Travis CI billing, the Dask org is migrating CI to GitHub Actions.

This repo contains a `.travis.yml` file which needs to be replaced with an equivalent `.github/workflows/ci.yml` file.

See dask/community#107 for more details.

### PR DESCRIPTION
**Pull request by jacobtomlinson**
Closes #6787

This PR migrates CI to GitHub Actions.

To do:
- [x] Run tests on all os/python combinations
- [x] Run nightly tests on upstream repos
- [x] Import checks
- [x] Mindeps checks
- [x] HDFS checks

Changes:
- Coverage is only run on linux py38
- All python versions are now tested on macos

### CODE REVIEW

**Comment by jacobtomlinson** [comment]
I *think* this is ready now. 

The mindeps and upstream tests are failing on my fork, but the failures are consistent with the existing failures in Travis. This should probably be investigated but is orthogonal to this migration.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,34 +1,46 @@
...
-      run: |
-        cd docs
-        make html
+      - name: Coverage
+        shell: bash -l {0}
+        if: ${{ matrix.os == 'ubuntu-latest' &&  matrix.python-version == '3.8' }}
+        env:
+          COVERAGE: "true"
+          PARALLEL: "false"
+        run: source continuous_integration/scripts/coverage.sh
```
Comment (jrbourbeau) [inline .github/workflows/ci.yml:46]: Currently we're getting a `No data to report.` error when this runs. We might consider switching over to using [Coverall's GitHub Action](https://github.com/marketplace/actions/coveralls-github-action). Additionally, while coverage is nice to have, I'd be okay commenting this out for now and addressing it in a follow-up PR if it means we can merge this sooner

File: .github/workflows/ci-upstream.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+name: Upstream
+
+on:
+  schedule:
+    - cron: "0 1 * * *"
+  push:
+  pull_request:
```
Comment (jrbourbeau) [inline .github/workflows/ci-upstream.yml:7]: Currently these CI builds will run on pushes to forks. We should consider only running on pushes / pull requests to the main branch.

File: .github/workflows/ci-upstream.yml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+jobs:
+  build:
+    runs-on: "ubuntu-latest"
+    if: "contains(github.event.head_commit.message, 'test-upstream') || github.event_name != 'pull_request'"
```
Comment (jrbourbeau) [inline .github/workflows/ci-upstream.yml:12]: I'm not 100% sure this is working as the upstream build was skipped for e5e2cd9872c3833897c8ec87bf905d2299bcd456 which includes "test-upstream" in the commit message

File: continuous_integration/travis/install.sh [hunk 4]
```diff
@@ -1,54 +0,0 @@
...
-conda config --set always_yes yes --set changeps1 no --set remote_max_retries 10
-
-# Create conda environment
-conda env create -q -n test-environment -f $ENV_FILE
-source activate test-environment
-
-# We don't have a conda-forge package for cityhash
-# We don't include it in the conda environment.yaml, since that may
-# make things harder for contributors that don't have a C++ compiler
-python -m pip install --no-deps cityhash
```
Comment (jrbourbeau) [inline continuous_integration/travis/install.sh:30]: Looks like we are no longer installing `cityhash` in the new `install.sh` script

File: .github/workflows/ci-additional.yml [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+          miniconda-version: "latest"
+          python-version: "3.7"
+          environment-file: continuous_integration/environment-${{ matrix.environment }}.yaml
+          activate-environment: test-environment
+          auto-activate-base: false
+
+      - name: Install
+        shell: bash -l {0}
+        env:
+          UPSTREAM_DEV: 1
```
Comment (jrbourbeau) [inline .github/workflows/ci-additional.yml:30]: Previously we weren't testing with `UPSTREAM_DEV` for the mindeps tests. As things can break with development versions of packages, we might consider not setting `UPSTREAM_DEV=1` here

**Review by jrbourbeau — COMMENTED** [review]
Thanks for all your work on this @jacobtomlinson! I've left a few small-ish comments, but overall the changes here look good

File: .github/workflows/ci-upstream.yml (hunk 3 above)
Comment (jacobtomlinson) [reply .github/workflows/ci-upstream.yml:12]: It only works for the first commit in the PR. I spent some time trying to figure out if you could trigger this on the latest commit on the branch but it doesn't seem the `github.event` object has this info.

**Comment by jacobtomlinson** [comment]
Thanks @jrbourbeau I've addressed the review feedback.

I've commented out the coverage for now. I would actually propose that we switch to codecov in future. I've had good experience in other projects and it handles combining coverage very well. So you run coverage on everywhere and codecov merges the reports, this would mean things like the HDFS tests would be included in coverage on master.

But as you say it would be good to address this later.

File: .github/workflows/ci-additional.yml [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    strategy:
+      matrix:
+        environment: ["mindeps-array-dataframe", "mindeps-bag-delayed"]
+
+    steps:
+      - name: Checkout source
+        uses: actions/checkout@v2
+
+      - name: Setup Conda Environment
+        uses: goanpeca/setup-miniconda@v1.7.0
```
Comment (gforsyth) [inline .github/workflows/ci-additional.yml:19]:
```suggestion
        uses: conda-incubator/setup-miniconda@v1.7.0
```

`goanpeca` migrated their action to the `conda-incubator` org

File: .github/workflows/ci-additional.yml [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    strategy:
+      fail-fast: true
+      matrix:
+        python-version: ["3.6", "3.7", "3.8"]
+    steps:
+      - name: Checkout source
+        uses: actions/checkout@v2
+
+      - name: Setup Conda
+        uses: goanpeca/setup-miniconda@v1.7.0
```
Comment (gforsyth) [inline .github/workflows/ci-additional.yml:66]:
```suggestion
        uses: conda-incubator/setup-miniconda@v1.7.0
```

File: .github/workflows/ci-upstream.yml [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+    steps:
+      - name: Checkout source
+        uses: actions/checkout@v2
+
+      - name: Setup Conda Environment
+        uses: goanpeca/setup-miniconda@v1.7.0
```
Comment (gforsyth) [inline .github/workflows/ci-upstream.yml:19]:
```suggestion
        uses: conda-incubator/setup-miniconda@v1.7.0
```

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -1,34 +1,43 @@
...
+    steps:
+      - name: Checkout source
+        uses: actions/checkout@v2
 
-    - name: Set up Python 3.7
-      uses: actions/setup-python@v1
-      with:
-        python-version: 3.7
+      - name: Setup Conda Environment
+        uses: goanpeca/setup-miniconda@v1.7.0
```
Comment (gforsyth) [inline .github/workflows/ci.yml:24]:
```suggestion
        uses: conda-incubator/setup-miniconda@v1.7.0
```

**Review by gforsyth — COMMENTED** [review]
Small changes for the migration from `goanpeca` -> `conda-incubator` org

**Review by jrbourbeau — APPROVED** [review]
Thanks for all your work on this @jacobtomlinson!

**Comment by jacobtomlinson** [comment]
Thanks for the tweaks @gforsyth!

**Pull request merged**